Promote experimental proxy-status command to regular, and remove regular `proxy-status` command

hanxiaop opened the pull request:
**Please provide a description of this PR:**
Plan to ~remove~ promote the `experimental proxy-status` command in the next release. This command mirrors the `proxy-status` command, necessitating duplicate efforts in maintenance whenever the `proxy-status`  command is updated.

There were some discussions related to this: https://github.com/istio/istio/pull/46050

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot-istiodremote_istio/1721457624412590080) | true | `/test integ-pilot-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@hanxiaop: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot-istiodremote_istio/1721457624412590080) | true | `/test integ-pilot-istiodremote`
integ-cni_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-cni_istio/1721457608902053888) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@hanxiaop: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot-istiodremote_istio/1721457624412590080) | true | `/test integ-pilot-istiodremote`
integ-cni_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-cni_istio/1721457608902053888) | true | `/test integ-cni`
integ-pilot_istio | ed67f9ab5ccee505341bc4503e8680162cc5b24a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot_istio/1721457621065535488) | true | `/test integ-pilot`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

GregHanson commented on tests/integration/pilot/piggyback_test.go:
```diff
@@ -64,37 +62,5 @@ func TestPiggyback(t *testing.T) {
-			}
-
-			// Test gRPC-based Tap Service using istioctl.
-			retry.UntilSuccessOrFail(t, func() error {
```
the other integration test was a duplicate of the base `istioctl ps`, but are we loosing coverage in this one by removing this test?  is it no longer valid or is it covered somewhere else?

GregHanson replied:
I think some of this was discussed here: https://github.com/istio/istio/pull/46050#issuecomment-1661953633

if this is a loss of functionality - do we want to remove the test, or do we want to create an issue for tracking and use a `t.Skip(<link to GitHub issue>)` ?

hanxiaop replied:
You are right, this will lead to less coverage I think. I think before doing `t.Skip` or creating an issue, I'll look into the code, and try to replace it with a new test method to test the gPRC-based Tap instead of `istioctl x ps`.

hanxiaop replied:
This should be addressed by https://github.com/istio/istio/pull/47801.

howardjohn commented:
I want to do some more investigating usage before merging. will report back soon

howardjohn commented:
also the point of the experimental command was to replace the old one. This does the opposite. What changed?

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-security-istiodremote_istio | 5ef6bd5ad7e591cb79b9275f7fd6a3deb1ecf959 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-security-istiodremote_istio/1723075146849193984) | unknown | `/test integ-security-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hanxiaop commented:
> also the point of the experimental command was to replace the old one. This does the opposite. What changed?

@howardjohn The intention is to avoid dup efforts in maintaining two commands `ps` and `experimental ps` since they are quite similar at this moment. If there are any concerns about removing the experimental command, we can also replace it with the experimental command, which provides more ways to get the proxy status. I'm looking forward to see the usage.

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

ingwonsong commented:
> also the point of the experimental command was to replace the old one.

big +1 to promoting the experimental one to regular, if it is necessary.

it is because:

1. There would be the reason why we had to implement the experimental one even though it was "re-inventing the wheel"
2. The experimental one provides various way to access the `proxy-status` information, which means much enhanced than the old one.

hanxiaop commented:
> 1. There would be the reason why we had to implement the experimental one even though it was "re-inventing the wheel"
> 2. The experimental one provides various way to access the `proxy-status` information, which means much enhanced than the old one.

IIRC the original intention was to replace the original `ps` command. Since `x ps` is also well-maintained, I think promoting it is appropriate. Currently, `ps` should be considered a subset of `x ps`.

istio-testing commented:
@hanxiaop: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-security-istiodremote_istio | 5ef6bd5ad7e591cb79b9275f7fd6a3deb1ecf959 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-security-istiodremote_istio/1723075146849193984) | unknown | `/test integ-security-istiodremote`
integ-pilot-istiodremote_istio | 4975471bb33b877aaa9ecdfb72adeeaf80906537 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot-istiodremote_istio/1726847482031247360) | true | `/test integ-pilot-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@hanxiaop: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-istiodremote_istio | 4975471bb33b877aaa9ecdfb72adeeaf80906537 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-pilot-istiodremote_istio/1726847482031247360) | true | `/test integ-pilot-istiodremote`
lint_istio | fd0eca2f671b77ff6151ef4a77cb1b67e28ccf8f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/lint_istio/1726857266205822976) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hanxiaop commented:
@GregHanson @howardjohn @ingwonsong Can any of you bring this to the WG meeting to gather consensus on the promotion? Thanks.

GregHanson commented:
@hanxiaop added to the agenda for next week's community meeting - since this week's meeting is cancelled due to US holiday this week

hanxiaop commented:
> @hanxiaop added to the agenda for next week's community meeting - since this week's meeting is cancelled due to US holiday this week

Sounds good, thanks

GregHanson commented:
discussion from 11/29 community call: ([notes](https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit?pli=1#heading=h.3lzzxgz0kjhk))

TLDR;
 - general approval for the promotion

TODO's
 - verify test coverage of `x ps`
 - `x ps` needs to remain for N releases before finally being removed
 - long term follow-up: open issue for tracking deprecation of the HTTP debug API on istiod
   - for backwards compatibility, older istioctl binaries can still be used to access the original `ps` command

GregHanson commented on releasenotes/notes/47740.yaml:
```diff
@@ -0,0 +1,6 @@
+area: istioctl
+releaseNotes:
+- |
+  **Updated** `istioctl experimental proxy-stauts` has been promoted to `istioctl proxy-status`. The old `istioctl proxy-status` command is removed.
```
Should probably clarify the release note a little more.  Add something along the lines of `This promotion should not result in any loss of functionality.  Adds additional flags to support multi-control plane environments`

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

hanxiaop commented:
@GregHanson I think this PR meets most of the TODO's, and I'll create an issue to track the HTTP debug API.

> verify test coverage of x ps

I think that currently, both `ps` and `x ps` are being tested by this PR. Although they are the same, we need to maintain `x ps` for several releases, so I prefer testing both which makes more sense. See the integration test changes for examples.

One question is: do we need to add tests for the xDS related flags? I couldn't find any existing tests for these options, which were introduced in https://github.com/istio/istio/pull/25087 without any tests. BTW there are already graduated commands like `istioctl kube-inject` that use these flags. Besides the flags, I think the command is covered with tests.

GregHanson reviewed (approved):
looks good to me, but I will wait for any last comments from @howardjohn before removing the `no-not-merge` label

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

linsun commented on releasenotes/notes/47740.yaml:
```diff
@@ -0,0 +1,8 @@
+area: istioctl
+releaseNotes:
+- |
+  **Updated** `istioctl experimental proxy-stauts` has been promoted to `istioctl proxy-status`. The old `istioctl proxy-status` command has been removed.
+  This promotion should not result in any loss of functionality. However, the request is now sent based on xDS instead of HTTP, and we have introduced a 
+  set of new xDS-based flags to target the control plane.
```
Could you explain the existing `istioctl ps` cmd?  To readers, it is confusing we have promoted the `istioctl x ps` while we already has `istioctl ps` today.

hanxiaop replied:
Great suggestion! I'll revise it.

linsun reviewed (commented):
RN needs some work

howardjohn commented:
`x ps` has ~3 modes:

1. Query each istiod and portforward
2. xds-via-agents
3. xds-address

(1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).

Do we want to promote all 3?

Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

hanxiaop commented:
> `x ps` has ~3 modes:
> 
> 1. Query each istiod and portforward
> 2. xds-via-agents
> 3. xds-address
> 
> (1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).
> 
> Do we want to promote all 3?
>
> Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

1 should not be a problem. I think 3 won't be either, as it is quite similar to 1 but applies directly to the xds-address. The xds-address has been tested and is already used in graduated commands like `kube-inject`. I agree there may be uncertainty about 2 and seems that disallowing some flags would be a great approach to gradually promote the command.

I prefer to hold off on this until 1.22 to allow more time for research and testing, which will help us gain confidence.

hanxiaop commented:
not stale

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | db1c74a4c0adc12b02203a801009f2fcdfb7c61c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/lint_istio/1764857868743348224) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hanxiaop commented:
> `x ps` has ~3 modes:
> 
> 1. Query each istiod and portforward
> 2. xds-via-agents
> 3. xds-address
> 
> (1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).
> 
> Do we want to promote all 3?
> 
> Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

@howardjohn I've marked option 2 as hidden in https://github.com/istio/istio/pull/47740/commits/7e1366c553f917c7b714c9c70ae143e9ef7847ed. For option 3, the flags already exist in graduated flags like `kube-inject`. Although the logic is not identical, I have tested it and verified that it can work as expected. I think if there are any bugs, we can do a forward fix. Let me know if this addresses your concerns.

howardjohn commented on istioctl/pkg/proxystatus/proxystatus_test.go:
```diff
@@ -46,8 +46,8 @@ type execTestCase struct {
-			args:           []string{},
-			expectedString: "NAME     CLUSTER     CDS     LDS     EDS     RDS     ECDS     ISTIOD",
+			args:          []string{},
+			wantException: true, // Error: no running Istio pods in "istio-system"
```
seems we are removing valid test coverage here?

hanxiaop replied:
The test is incorrect because the fake client returns an empty response with no error. In a real cluster, errors are shown for both commands when there is no Istiod pod running.
```
➜  istio git:(master)  istioctl ps                                 
Error: unable to find any Istiod instances
➜  istio git:(master) istioctl x ps
Error: no running Istio pods in "istio-system"
```

howardjohn replied:
every test case now gives errors, can we have a test that covers success?

hanxiaop replied:
Added, thanks!

howardjohn reviewed (approved):
LGTM just one test comment. We should merge https://github.com/istio/istio/pull/49728 too

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 5073197fcf6063c1f0ad51fc1b42f2cbea893218 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/gencheck_istio/1765269802097577984) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@hanxiaop: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 5073197fcf6063c1f0ad51fc1b42f2cbea893218 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/gencheck_istio/1765269802097577984) | true | `/test gencheck`
lint_istio | 5073197fcf6063c1f0ad51fc1b42f2cbea893218 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/lint_istio/1765269812088410112) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-telemetry-istiodremote_istio | f43791a57710747cb6d53912b2aed95dc374e20e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-telemetry-istiodremote_istio/1766001992313868288) | unknown | `/test integ-telemetry-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hanxiaop commented:
/test integ-telemetry-istiodremote

istio-testing commented:
@hanxiaop: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-telemetry-istiodremote_istio | f43791a57710747cb6d53912b2aed95dc374e20e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/47740/integ-telemetry-istiodremote_istio/1766009702069571584) | true | `/test integ-telemetry-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
